pkg/csr/csr.go

```diff
@@ -0,0 +1,236 @@
+var kubeletClientUsages = []certificatesv1.KeyUsage{
+	certificatesv1.UsageKeyEncipherment,
+	certificatesv1.UsageDigitalSignature,
+	certificatesv1.UsageClientAuth,
+}
+
+// kubeletServerUsages contains the permitted key usages from a kubelet-serving signer
+var kubeletServerUsages = []certificatesv1.KeyUsage{
+	certificatesv1.UsageKeyEncipherment,
+	certificatesv1.UsageDigitalSignature,
+	certificatesv1.UsageServerAuth,
+}
+
+const (
+	nodeGroup      = "system:nodes"
+	nodeUser       = "system:node"
+	NodeUserPrefix = nodeUser + ":"
+	systemPrefix   = "system:authenticated"
+)
+
+// CSRApprover holds the information required to approve a node CSR
+type CSRApprover struct {
+	k8sClientSet      *kubernetes.Clientset
+	instanceAddresses []string
+	log               logr.Logger
+}
+
+// NewCSRApprover returns a pointer to the CSRApprover
+func NewCSRApprover(k8sclientSet *kubernetes.Clientset, instanceAddresses []string, log logr.Logger) (*CSRApprover, error) {
+	if len(instanceAddresses) == 0 || k8sclientSet == nil {
+		return nil, errors.New("instance address list or kubernetes clientset should not be nil")
+	}
+	return &CSRApprover{k8sclientSet, instanceAddresses, log}, nil
+}
+
+// validateCSRContents returns true if the CSR request contents are valid
+func (c *CSRApprover) validateCSRContents(csr *certificatesv1.CertificateSigningRequest) (bool, error) {
+	parsedCSR, err := ParseCSR(csr.Spec.Request)
+	if err != nil {
+		return false, errors.Wrapf(err, "error parsing CSR %s", csr.Name)
+	}
+
+	dnsAddr := strings.TrimPrefix(parsedCSR.Subject.CommonName, NodeUserPrefix)
+	if dnsAddr == "" {
+		return false, errors.Errorf("CSR subject name does not contain the required node user prefix %s", csr.Name)
+	}
+
+	// check DNS of node requesting CSR matches instance IP/DNS address
+	var hasMatch bool
+	for _, instanceAdd := range c.instanceAddresses {
+		if areAddressesEqual(instanceAdd, dnsAddr) {
+			hasMatch = true
+			break
+		}
+	}
+	if !hasMatch {
+		// CSR is not from a BYOH Windows instance, don't deny it or return error
+		// since it might be from a linux node.
+		return false, nil
+	}
+	// Kubelet on a node needs two certificates for its normal operation:
+	// Client certificate for securely communicating with the Kubernetes API server
+	// Server certificate for use by Kubernetes API server to talk back to kubelet
+	// Both types are validated based on their contents
+	var validated bool
+	if isNodeClientCert(csr, parsedCSR) {
+		// Node client bootstrapper CSR is received before the instance becomes a node
+		// hence we should not proceed if a corresponding node already exists
+		_, err := c.k8sClientSet.CoreV1().Nodes().Get(context.TODO(), dnsAddr, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, errors.Wrapf(err, "unable to get node %s", dnsAddr)
+		} else if err == nil {
+			return false, errors.Wrapf(err, "%s node already exists, cannot validate CSR", dnsAddr)
```

Comment:
Is `dnsAddr` equivalent to the `node.Name`? Because that is what this error is hinting at. 

Is this how the `cluster-machine-approver` is handling this scenario?

Reply:
I think the naming caused some confusion here, it should be named nodeName. The cluster-machine-appover handles this by matching the nodeName with the machine.status.address && type internalDNSAddress.